apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/parser/GraphQLDocumentParser.kt

```diff
@@ -645,29 +645,48 @@ class GraphQLDocumentParser(val schema: Schema, private val packageNameProvider:
     }
   }
 
-  private fun Set<String>.usedSchemaTypes(exclude: Set<String> = emptySet()): Set<Schema.Type> {
+  private fun Set<String>.usedSchemaTypes(): Set<Schema.Type> {
     if (isEmpty()) {
       return emptySet()
     }
 
-    val usedSchemaTypes = filter { ScalarType.forName(it) == null }
+    val (scalarTypes, inputObjectTypes) = filter { ScalarType.forName(it) == null }
         .map { schema[it] ?: throw GraphQLParseException(message = "Undefined schema type `$it`") }
         .filter { type -> type.kind == Schema.Kind.SCALAR || type.kind == Schema.Kind.ENUM || type.kind == Schema.Kind.INPUT_OBJECT }
-        .toSet()
+        .partition { type -> type.kind == Schema.Kind.SCALAR || type.kind == Schema.Kind.ENUM }
+        .let { (scalarTypes, inputObjectTypes) ->
+          @Suppress("UNCHECKED_CAST")
+          scalarTypes to (inputObjectTypes as List<Schema.Type.InputObject>)
+        }
 
-    val inputObjectUsedTypes = usedSchemaTypes
-        .mapNotNull { type -> type as? Schema.Type.InputObject }
-        .flatMap { inputObject -> inputObject.usedTypes(exclude = this + exclude) }
-        .toSet()
-        .usedSchemaTypes(this + exclude)
+    val usedTypes = (scalarTypes + inputObjectTypes).toMutableSet()
+    val visitedTypeNames = scalarTypes.map { it.name }.toMutableSet()
 
-    return usedSchemaTypes + inputObjectUsedTypes
-  }
+    val inputTypesToVisit = inputObjectTypes.toMutableList()
+    while (inputTypesToVisit.isNotEmpty()) {
+      val inputType = inputTypesToVisit.removeAt(inputTypesToVisit.lastIndex).also {
+        usedTypes.add(it)
+        visitedTypeNames.add(it.name)
+      }
+      val (nestedScalarTypes, nestedInputTypes) = inputType
+          .inputFields
+          .asSequence()
```

Comment:
Just curious: does this actually save time ?

Reply:
No, it's just a small optimization probably even not noticeable.

What really saves time is:
1) recursion replaced with loop
2) we were visiting the same input types over and over when traverse deep nested tree.